### ISSUE TITLE
Multiple hashes in osVersionRequirements results in undefined behavior

### ISSUE THREAD

**Issue by erikng** [macadmins/nudge#157]
This is somewhat similar to https://github.com/macadmins/nudge/issues/40 in that it will need major re-architecture changes to resolve.

So I found a bug with Nudge, but I'm not sure if anyone is even using Nudge in this way, and I'm debating dropping it entirely unless I can understand the benefit.

In the beginning @groob had mentioned creating a list of OS enforcements and having different rules for them, but thinking through it, I'm not sure if I like the complexity.

Let's say you are on `11.2.1` and you set one nudge event for `11.2.2` and another event for `11.2.3`. Which should we respect?
Is anyone in here actually creating multiple nudge event windows right now? I just tried it and it doesn't work. Nudge currently loops through the list and when it finds a nudge event in scope, it uses that for its logic.

What kind of use cases exist for dual nudge events? Perhaps I'm not thinking through it. I was curious to finally try it, but now that I realize it's broken, I can't imagine anyone has even used it.

```json
{
  "osVersionRequirements": [
    {
      "requiredInstallationDate": "2021-03-29T00:00:00Z",
      "requiredMinimumOSVersion": "11.2.2",
      "targetedOSVersions": [
        "11.0",
        "11.0.1",
        "11.1",
        "11.2",
        "11.2.1"
      ]
    },
    {
      "requiredInstallationDate": "2021-03-15T00:00:00Z",
      "requiredMinimumOSVersion": "11.3",
      "targetedOSVersions": [
        "11.0",
        "11.0.1",
        "11.1",
        "11.2",
        "11.2.1",
        "11.2.2",
        "11.2.3"
      ]
    }
  ],
}
```

So in the example above, we have two different kinds of failures

- If you're on `11.2.3`, since you meet or exceed the initial `requiredMinimumOSVersion`, you don't get nudged - the second event doesn't matter
- If you're on `11.2.1`, you get the first Nudge event, since you're in scope.

I keep trying to think of why you would want two nudge events and can't really think why it would make sense, unless you you did something like this.

```json
{
  "osVersionRequirements": [
    {
      "requiredInstallationDate": "2021-03-15T00:00:00Z",
      "requiredMinimumOSVersion": "11.1",
      "targetedOSVersions": [
        "11.0",
        "11.0.1"
      ]
    },
    {
      "requiredInstallationDate": "2021-03-17T00:00:00Z",
      "requiredMinimumOSVersion": "11.2.3",
      "targetedOSVersions": [
        "11.1",
        "11.2",
        "11.2.1",
        "11.2.2"
      ]
    }
  ],
}
```

But in a use case like this (say if it was March 18th), wouldn't you really just want everyone on 11.2.3? You'd potentially punish users who are on 11.0/11.0.1, get on 11.1, only to then get nudged immediately when they login post 11.1.

**Comment by erikng** [comment]
Which hash should we respect if you are in scope for two of them?
```json
{
  "osVersionRequirements": {
    "requiredInstallationDate": "2021-03-17T00:00:00Z",
    "requiredMinimumOSVersion": "11.2.1",
    "targetedOSVersions": [
      "11.0"
    ]
  },
  {
    "requiredInstallationDate": "2021-03-19T00:00:00Z",
    "requiredMinimumOSVersion": "11.2.2",
    "targetedOSVersions": [
      "11.0",
      "11.0.1"
    ]
  },
}
```
What should be the expected behavior for poor little 11.0 machine?

This feels very muddy and full of unexpected behaviors.

**Comment by erikng** [comment]
@natewalck 
> I’d go with the most recent version but an admin should change it as well. When I cut the pr to add a 11.2.2 block, I’d also update the first block

and then you'd hit the bug I just hit :disappointed:
two bugs

- If you are in scope for two hashes, only the first hash is considered
- if you are running an OS that's higher than the minimumOS of the first hash, only the first hash is considered

**Comment by erikng** [comment]
@groob 
> are you considering multiple supported major versions in your test cases? ex 10.15.x, 11 and 12. And the window where users can run either 11.x and 12.x (for example)

> this is a problem for me I’d like to eliminate (it punishes the good users more, and doesn't really address the bad users)

no! that was the use case I couldn't think of!

**Comment by erikng** [comment]
I think we need `NSUserDefaults` to save nudge enforcement data to fix this issue and also work on the idea of [user deferrals](https://github.com/macadmins/nudge/issues/40).
If say a user deferred a bunch of 11.2.1, then you start nudging 11.2.2, if they never completed their last nudge, they shouldn't get a brand new set of nudges but combine them. But this logic is going to get very complex very quickly.


@groob 
> i’ve been debating doing that vs setting a obsolete parameter in the config and then being able to say anything below N is now past due, anything after follows the regular reminder policy

that's an interesting idea. so more like `requiredMinimumOSVersionThreshold` vs `requiredMinimumOSVersion`

@groob 
> a new key, but basically yes. it would allow setting 11.2.3 as the threshold while I start nudging for 11.3.x for the next month and get everyone below to remain on more aggressive notifications than users who are now updating more often

**Comment by smithjw** [comment]
@erikng Have you though about this feature-request since March? In our case we require devices to be on a specific version or higher if there are known vulns in the wild. We currently manage this with two LaunchAgents which look at two separate local json files. Within each of the files, we have exclusive `requiredInstallationDate`,  & `targetedOSVersions` (in that they do not clash)

Standard:
```
"requiredInstallationDate": "2021-09-17T11:59:00Z",
"requiredMinimumOSVersion": "11.5.2",
"targetedOSVersions": ["11.5.1"]
```

Critical:
```
"requiredInstallationDate": "2021-08-12T01:59:00Z",
"requiredMinimumOSVersion": "11.5.2",
"targetedOSVersions": ["11.0","11.0.1","11.1","11.2","11.2.1","11.2.2","11.2.3","11.3","11.3.1","11.4","11.5"]
```

In the above case our absolute minimum version to not Nudge people hourly is to be on 11.5.1 (it states 11.5.2 in the Critical json because that is what is released right now; so mainly for cosmetic reasons). If someone is on a version less than 11.5.1, Software Update will take them straight to 11.5.2 rather than 11.5.1.

Now that we have User Deferrals, the way I would think about the logic is as follows:

- Deploy a single LaunchAgent that fires every hour
- Allow the inclusion of multiple `osVersionRequirements` dicts
- Match the Mac against the first `osVersionRequirements` that the given Mac falls into
- If the Mac is on a higher `requiredMinimumOSVersion` than specified in the first dict, check the next dict
- Once it falls into scope for one `osVersionRequirements`, stop matching against others (it's the responsibility of the Admin to make sure `targetedOSVersions` do not double up
- If `deferRunUntil` is set to after the `requiredInstallationDate` for a given dict that the Mac falls into, it should be ignored/reset and the Mac should be Nudged based on the LaunchAgent/prefs

**Comment by erikng** [comment]
I'll be honest. That sounds like a lot of work for your specific use case. We have uses cases like this as well, but use configuration management (chef) to dynamically change the properties. 

I can possibly try and fix the bug, but I didn't plan on it for v1.1.0. As far as your other requests, I don't think I'm willing to create all the logic flows and test them.

**Comment by erikng** [comment]
Actually, what you describe is likely what I would have to implement for this to all work :(

**Comment by smithjw** [comment]
All good, just wanted to document the specific use-case here hoping it would justify the effort required a bit more 😄 . As it stands, I already have a workaround in the dual LaunchAgents+Configs that works.

Would you suggest running something like `defaults delete  ~/Library/Preferences/com.github.macadmins.Nudge.plist deferRunUntil` when deploying new OS target versions?

**Comment by erikng** [comment]
When you deploy new targetOS versions, the defer key is ignored and reset (as are the deferrals), so long as the new target is higher than the old target.

**Comment by smithjw** [comment]
Found a "bug" with the new deferrals. I say "bug" as it's based on my use case and having two LaunchAgents

Running Standard Config
```
"Device pastRequiredInstallationDate: false"
"User has selected a deferral date (2021-08-20 02:14:39 +0000) that is greater than the launch date (2021-08-20 01:15:33 +0000)"
"Nudge is terminating due to condition met"
```
☝️ This is expected as I've deferred for an hour and not past the deadline

Running Critical Config
```
"Device pastRequiredInstallationDate: true"
"User has selected a deferral date (2021-08-20 02:14:39 +0000) that is greater than the launch date (2021-08-20 01:15:36 +0000)"
"Nudge is terminating due to condition met"
```
☝️ Not expected as I'm past the install deadline

If I run `defaults delete ~/Library/Preferences/com.github.macadmins.Nudge.plist deferRunUntil`, the Critical Nudge session launches as expected not showing any way to dismiss the window as I'm passed the install date.

I don't expect this to be addressed as it's a real edge-case, just highlighting

**Comment by erikng** [comment]
This is a bug I want to fix though it isn't something I could have accounted for. The defer button isn't supposed to allow the user to defer past the deadline, but it sounds like in this case you have bumped up the deadline. That should definitely be solved.

in this case are you also bumping the min OS version or just the install date?

**Comment by smithjw** [comment]
> The defer button isn't supposed to allow the user to defer past the deadline
I deferred Nudge when launched via a preference file that listed a `requiredInstallationDate` in the future

> but it sounds like in this case you have bumped up the deadline
I then launched Nudge again with a different preference file that listed the `requiredInstallationDate` in the past.

In both preference files, the `requiredMinimumOSVersion` was listed as the same OS version

**Comment by erikng** [comment]
Okay that's what I figured. This is a cool bug, that I think I can solve quite easily. 

Can you put this in another GitHub issue so I can properly track it? It doesn't really have anything to do with this ticket.

**Comment by groob** [comment]
The examples above all revolve around macOS 11. When I was suggesting multiple versions, I only meant multiple major versions, and a single minor/patch for each major version. Basically, if you have a large fleet where both macOS 11 and macOS 12 are supported, you can specify a minor version for each. That way, the client which is running macOS 11 would ignore the macOS 12 build, and notify the user about 11.5.2 or whatever patch you were pointing at. And clients running macOS 12, would be notifying about macOS 12.1.1. 

In the future, when macOS 11 becomes unsupported, you simply remove it from the list, and all the clients begin alerting about macOS 12.

**Comment by groob** [comment]
Why is there a targeted OS Versions list, rather than assuming that everything higher than 11.0 and lower than 11.5.2 is targeted?

**Comment by erikng** [comment]
I anticipated a minor version coming out like 11.5.2 that didn't ultimately matter to the security of your fleet, so I wanted to be able to enforce it as the new minimum OS but not punish people who were secure (11.5.1 users).

In hindsight, you would probably just make 11.5.1 your required version and softwareupdate client would take it from there. 

I could likely drop this key, but I don't fully understand how the logic would work for the dual OS environment. Would required OS be an array and nudge handles the logic if there are two major OS versions?

**Comment by groob** [comment]
Nudge would detect your major OS version, and then find the requirements you've listed for that major version. 
If it doesn't find requirements for the current one (because you deprecate it and remove the requirements) it will alert the user until they upgrade to either of the supported versions. 

In other words if you have a list like 
```
[{
require: 12.1
...
},
{
require: 11.5.2
...
}]
```

A host running macOS 12 would alert on your settings for macOS 12. 
A host running macOS 11 would alert on your settings for macOS 11. 
A host running OS X 10.15 would nag the user to pick an upgrade.  

Nudge has a diverse community of user, and I'm not sure my simple logic would fit for everyone. But it works for me.

**Comment by erikng** [comment]
> A host running OS X 10.15 would nag the user to pick an upgrade.

If you do want to end up using Nudge, I would ask for help on building this. Accounting for this seems full of user behaviors and actions that to date, we've never accounted for.

Given that Nudge does not currently support 3 Operating Systems (it's 11/12 only right now), I think this feature request should go into another github issue and tracked separately.

> A host running macOS 12 would alert on your settings for macOS 12. A host running macOS 11 would alert on your settings for macOS 11.

As far as that logic goes, it would likely be that I would change the preference key structure to something like this

old
```
"requiredInstallationDate": "2021-08-12T01:59:00Z",
"requiredMinimumOSVersion": "11.5.2",
"targetedOSVersions": ["11.0","11.0.1","11.1","11.2","11.2.1","11.2.2","11.2.3","11.3","11.3.1","11.4","11.5"]
```

new
```
"requiredInstallationDate": "2021-08-12T01:59:00Z",
"requiredMinimumOSVersion": "["11.5.1", 12.0.1"]
```

`targetedOSVersions` would go away entirely.
- `osVersionRequirements` would be reduced to a single dictionary instead of an array of dictionaries
- If you specify a single value for `requiredMinimumOSVersion`, it will apply for all devices. 
- If you specify multiple values for `requiredMinimumOSVersion`, nudge will look at the major version and return the first version that matches and use that logic.

The drawback to this approach is you can't specify multiple `requiredInstallationDate` keys. To accommodate something like that, an admin would have to use the multiple JSON idea @smithjw currently uses (or use configuration management to dynamically set the keys). MDM only users could create multiple profiles and deploy via conditions.

I'm personally in favor of this as it reduces the complexity of the code for me and punts the work to the admin who has more complex needs than 98% of the admins.

**Comment by erikng** [comment]
@natewalck would like your feedback on this. I kind of feel since this would be a breaking change, that I want it in v1.1.0 now, so people don't get tricked up.

Given that Monterey is close to release, it would make sense to have this type of change ready to go.

**Comment by groob** [comment]
> The drawback to this approach is you can't specify multiple requiredInstallationDate keys

What's the advantage vs an array of dictionaries? IMO there's a significant drawback to requiring config management/MDM to template the setting, when the client could easily decide which dictionary to pick from. 

As far as major OS version deprecation, I'd be up for helping with this. We should open a different thread to discuss what that will look like.

**Comment by natewalck** [comment]
I think having a key to match on for the ruleset makes a lot of sense (Similar to what @groob mentioned). 

For instance:

```
"updateRules": {
    "10": {
      "requiredMajorOSVersion": "12",
      "requiredInstallationDate": "2021-08-12T01:59:00Z",
    },
   "11": {
      "requiredMinimumOSVersion": "11.5.1",
      "requiredInstallationDate": "2021-08-13T01:59:00Z",
    },
   "12": {
      "requiredMinimumOSVersion": "12.0.1",
     "requiredInstallationDate": "2021-08-15T01:59:00Z",
    },
  } 
}
```

This could then be used for subversions too:

```
"updateRules": {
    "10": {
      "requiredMinimumOSVersion": "12.0",
      "requiredInstallationDate": "2021-08-12T01:59:00Z",
    },
   "11.4": {
      "requiredMinimumOSVersion": "11.5.2",
      "requiredInstallationDate": "2021-08-10T01:59:00Z",
    },
   "11.5.1": {
      "requiredMinimumOSVersion": "11.5.2",
      "requiredInstallationDate": "2021-08-24T01:59:00Z",
    },
   "12": {
      "requiredMinimumOSVersion": "12.0.1",
     "requiredInstallationDate": "2021-08-15T01:59:00Z",
    },
  } 
}
```

This would simplify the nudge code as you'd match against the version in the update rules dictionary and load the ruleset for that version. Then, you can have two rulesets, for instance going from 11.4 => 11.5.2 is more urgent than going 11.5.1 => 11.5.2.

For matching against the ruleset, it would be pretty straightforward too, in really janky pseudo-code:

```
## This should be a recursive function that either returns a ruleset or nil
curVersion = 11.4
rules = updateRules[curVersion]
if rules == nil {
  rules = rules.split('.')[-1]
}

rules = updateRules[curVersion]


```

Basically you try to find a ruleset for the full version, if it doesn't exist, lop off the x.x.N number and try again. If no match, lop off x.N and try again. If you can't match on a major version, its a configuration issue.

There could be more fun here depending how silly the admin provided `updateRules` are, but it allows for more specific targeting of sub-versions of a given OS.

**Comment by groob** [comment]
Ooh, I like that @natewalck!

At first I thought you're duplicating the major version for no reason, but I realize you get a lot of flexibility if you want. To target a specific OS version with a custom rule. Maybe have a `notes` field or something to customize the alert for just that specific version.

**Comment by 1dustindavis** [comment]
I also like this concept from an admin perspective. Lots of flexibility, while still being easy to parse. 

What would the client do if it did not find any matches in the config? For example, only 11.x and 12.x are in the config, but a 10.x machine tries to run it. Maybe a "default" or "fallback" option to catch any OS that is not defined?

**Comment by erikng** [comment]
I like @natewalck 's idea.

so the rules would be
- match on MAJOR.MINOR.PATCH and return that condition
- If no match, check MAJOR.MINOR and return that condition
- If no match, check MAJOR and return that condition
- If "default" value is present in the array, return that. 

I'm not sure if I would agree that it would "simplify" the nudge code base at all, but I do agree that it makes sense to change osVersionRequirements to this.

Dynamically changing the nudge UI text for this new behavior is out of scope and not something I'm sure I want to currently do as it would essentially mean changing the entire way the userInterface key is handled. I've been thinking instead about being able to pass a set of look-up values, say {osVersion} and nudge parsing that into the UI, so that could potentially handle this use case if designed that way. I'd recommend making another issue for that part as well.

**Comment by natewalck** [comment]
> I'm not sure if I would agree that it would "simplify" the nudge code base at all, but I do agree that it makes sense to change osVersionRequirements to this.

What I meant was if the config is an array of required OSes only, you'd have to do a fair amount of logic on the back end to make it work well (compare all items in array vs current and do the right thing). With a more explicit config, we can use a single recursive function to get the appropriate config back.

I agree with the default value, that would be a nice catch all.

**Comment by erikng** [comment]
I think I have a way for this to not be a breaking change from 1.0 to 1.1 upgrade.

This is what a real world v1.0 would look like
```
"osVersionRequirements": [
  {
    "majorUpgradeAppPath": "/Applications/Install macOS Monterey.app",
    "requiredInstallationDate": "2021-08-28T00:00:00Z",
    "requiredMinimumOSVersion": "11.5.2",
    "targetedOSVersions": [
      "11.0",
      "11.0.1",
      "11.1",
      "11.2",
      "11.2.1",
      "11.2.2",
      "11.2.3",
      "11.3",
      "11.3.1",
      "11.4",
      "11.5",
      "11.5.1"
    ]
  }
]
```

This is what a v1.1 would look like

```
"osVersionRequirements": [
  {
    "additionalRules": {
       "11": {
        "requiredMinimumOSVersion": "11.5.2",
        "requiredInstallationDate": "2021-08-10T01:59:00Z",
      },
      "11.5.1": {
        "requiredMinimumOSVersion": "11.5.2",
        "requiredInstallationDate": "2021-08-24T01:59:00Z",
      },
      "12": {
        "requiredMinimumOSVersion": "12.0.1",
       "requiredInstallationDate": "2021-08-15T01:59:00Z",
      },
    },
    "majorUpgradeAppPath": "/Applications/Install macOS Monterey.app",
    "requiredInstallationDate": "2021-08-28T00:00:00Z",
    "requiredMinimumOSVersion": "11.5.2",
    "targetedOSVersions": [
      "11.0",
      "11.0.1",
      "11.1",
      "11.2",
      "11.2.1",
      "11.2.2",
      "11.2.3",
      "11.3",
      "11.3.1",
      "11.4",
      "11.5",
      "11.5.1"
    ]
  }
]
```
In this case `targetedOSVersions` would be ignored in v1.1.0 and `additionalRules` would be ignored in v1.0. The original `requiredMinimumOSVersion` and `requiredInstallationDate` would be the "default/fallback" options, and only used if the `additionalRules` key doesn't exist in the JSON/Profile or any of the rules are not triggered.

I also think `majorUpgradeAppPath` needs to be another valued passed to `additionalRules`. In 2022, you may want Big Sur machines to be on macOS 13, but Monterey machines to stay on Monterey.

I would have to keep `osVersionRequirements` an array of dictionaries and still warn/ignore if the array has more than 1 element (ignoring all other elements besides the first one), but that would at least make upgrading to v1.1.0 a lot easier. I also think with this design it will fix @smithjw 's conundrum of critical/standard.

**Comment by erikng** [comment]
If anyone disagrees with ^, please speak up, otherwise I plan on trying to write this tomorrow.

**Comment by 1dustindavis** [comment]
Wouldn't this still be a breaking change? If someone deployed v1.1.0 without updating their config, it would break the functionality they expected.

If you really don't want a breaking change, I think you would need to maintain the existing behavior for targetedOSVersions, instead of ignoring it. 

(To be clear, I support breaking changes :)

**Comment by 1dustindavis** [comment]
I also don't love the name additionalRules, but it's fine.

**Comment by erikng** [comment]
> Wouldn't this still be a breaking change? If someone deployed v1.1.0 without updating their config, it would break the functionality they expected.
> 
> If you really don't want a breaking change, I think you would need to maintain the existing behavior for targetedOSVersions, instead of ignoring it.
> 
> (To be clear, I support breaking changes :)

Nope because nudge would effectively hit the "fallback".

**Comment by erikng** [comment]
> I also don't love the name additionalRules, but it's fine.

customBehaviors? fluffyBunnyDictionary?

**Comment by 1dustindavis** [comment]
> Nope because nudge would effectively hit the "fallback".

Oh, good point!

> customBehaviors? fluffyBunnyDictionary?

The more I think about it, additionalRules is not that bad.

**Comment by natewalck** [comment]
We are kind of stuck with the existing key structure (can't rename the keys to `requiredInstallationDate`, etc) so I think this format will work well while retaining compatibility. Perhaps `versionRules` for the new key?

**Comment by smithjw** [comment]
> customBehaviors? fluffyBunnyDictionary?

`osRules`?

**Comment by erikng** [comment]
ugh, I was looking into this, and remembering why `updateElements` is an array of dictionaries and why I used `_language` to work around it.

https://www.fivestars.blog/articles/codable-swift-dictionaries/

I'm too stupid to figure out dictionaries of dictionaries that support both JSON/mobileconfig via codable swift.

so now I'm sadly thinking

```
"osVersionRequirements": [
  {
    "versionRules": [
       {
        "majorUpgradeAppPath": "/Applications/Install macOS Monterey.app",
        "rule": "10",
        "requiredMinimumOSVersion": "12.0",
        "requiredInstallationDate": "2021-08-10T01:59:00Z"
      },
       {
        "majorUpgradeAppPath": "/Applications/Install macOS Monterey.app",
        "rule": "11",
        "requiredMinimumOSVersion": "11.5.2",
        "requiredInstallationDate": "2021-08-10T01:59:00Z"
      },
      {
        "rule": "11.5.1",
        "requiredMinimumOSVersion": "11.5.2",
        "requiredInstallationDate": "2021-08-24T01:59:00Z"
      },
      {
        "rule: "12",
        "requiredMinimumOSVersion": "12.0.1",
       "requiredInstallationDate": "2021-08-15T01:59:00Z"
      }
    ],
    "majorUpgradeAppPath": "/Applications/Install macOS Monterey.app",
    "requiredInstallationDate": "2021-08-28T00:00:00Z",
    "requiredMinimumOSVersion": "11.5.2",
    "targetedOSVersions": [
      "11.0",
      "11.0.1",
      "11.1",
      "11.2",
      "11.2.1",
      "11.2.2",
      "11.2.3",
      "11.3",
      "11.3.1",
      "11.4",
      "11.5",
      "11.5.1"
    ]
  }
]
```

The problem I already foresee is hitting the exact same issue I hit with this entire issue. Because it's an array of dictionaries, I will need to figure out if there is a way to mutate this out of codable and into something like `var fixedDict: [String : [String : String]]` so we can properly loop through it.

I really hate how Swift does this https://forums.swift.org/t/how-to-encode-objects-of-unknown-type/12253/10. Perhaps someone smarter can chime in.

**Comment by erikng** [comment]
I have it all working on https://github.com/macadmins/nudge/pull/224. Please check it out.

**Comment by groob** [comment]
Could you opt in to the new config behind a flag? And then in a few releases make it default?

**Comment by smithjw** [comment]
@groob is that required though? Erik mentioned that if you don't have the `additionalRules` dictionary in your config, Nudge would fall back to the old OS matching.

**Comment by groob** [comment]
It's just ugly to have the new config under a key like that. Perhaps instead of nesting it under `osVersionRequirements`, you could create `osVersionRequirements` and `osVersionRequirements2` or find a more creative name. It becomes opt in, with the ability to deprecate the old config in the next version, and leaving things at the top level.

**Comment by erikng** [comment]
That's a significant amount of work to create multiple codable keys, the logic around it, a feature flag, etc. 

Ugly or not, there's 0 chance I'm going to do all of that work.

If people don't care about it being a breaking change, then I can restructure the keys.

**Comment by erikng** [comment]
So we have completely overthought it. I have an idea I'm going to try and will post my results here in a few.

If I'm right, there will be no breaking changes, and I can actually solve this initial issue.

**Comment by erikng** [comment]
I'm much more happy with: https://github.com/macadmins/nudge/pull/225 . Please check it out.

**Comment by erikng** [comment]
@natewalck @groob and I are good about this so, going to close.

Documentation will be key and tracked at https://github.com/macadmins/nudge/issues/217
